server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/PersistLiveMeasuresStep.java

```diff
@@ -0,0 +1,133 @@
+
+public class PersistLiveMeasuresStep implements ComputationStep {
+
+  /**
+   * List of metrics that should not be persisted on file measure (Waiting for SONAR-6688 to be implemented)
+   */
+  private static final Set<String> NOT_TO_PERSIST_ON_FILE_METRIC_KEYS = unmodifiableSet(new HashSet<>(Arrays.asList(
+    FILE_COMPLEXITY_DISTRIBUTION_KEY,
+    FUNCTION_COMPLEXITY_DISTRIBUTION_KEY,
+    CLASS_COMPLEXITY_DISTRIBUTION_KEY)));
+
+  private final DbClient dbClient;
+  private final MetricRepository metricRepository;
+  private final MeasureToMeasureDto measureToMeasureDto;
+  private final TreeRootHolder treeRootHolder;
+  private final MeasureRepository measureRepository;
+
+  public PersistLiveMeasuresStep(DbClient dbClient, MetricRepository metricRepository, MeasureToMeasureDto measureToMeasureDto,
+    TreeRootHolder treeRootHolder, MeasureRepository measureRepository) {
+    this.dbClient = dbClient;
+    this.metricRepository = metricRepository;
+    this.measureToMeasureDto = measureToMeasureDto;
+    this.treeRootHolder = treeRootHolder;
+    this.measureRepository = measureRepository;
+  }
+
+  @Override
+  public String getDescription() {
+    return "Persist live measures";
+  }
+
+  @Override
+  public void execute() {
+    try (DbSession dbSession = dbClient.openSession(true)) {
+      dbClient.liveMeasureDao().deleteByProjectUuid(dbSession, treeRootHolder.getRoot().getUuid());
+      new DepthTraversalTypeAwareCrawler(new MeasureVisitor(dbSession)).visit(treeRootHolder.getRoot());
+      dbSession.commit();
+    }
+  }
+
+  private class MeasureVisitor extends TypeAwareVisitorAdapter {
+    private final DbSession session;
+
+    private MeasureVisitor(DbSession session) {
+      super(CrawlerDepthLimit.LEAVES, PRE_ORDER);
+      this.session = session;
+    }
+
+    @Override
+    public void visitAny(Component component) {
+      Multimap<String, Measure> measures = measureRepository.getRawMeasures(component);
+      persistMeasures(component, measures);
+    }
+
+    private void persistMeasures(Component component, Multimap<String, Measure> batchReportMeasures) {
+      for (Map.Entry<String, Collection<Measure>> measures : batchReportMeasures.asMap().entrySet()) {
+        String metricKey = measures.getKey();
+        if (NOT_TO_PERSIST_ON_FILE_METRIC_KEYS.contains(metricKey) && component.getType() == Component.Type.FILE) {
+          continue;
+        }
+
+        Metric metric = metricRepository.getByKey(metricKey);
+        LiveMeasureDao dao = dbClient.liveMeasureDao();
+        for (Measure measure : from(measures.getValue()).filter(NonEmptyMeasure.INSTANCE)) {
+          LiveMeasureDto dto = measureToMeasureDto.toLiveMeasureDto(measure, metric, component);
+          dao.insert(session, dto);
+        }
+      }
+    }
+
+  }
+
+  private enum NonEmptyMeasure implements Predicate<Measure> {
+    INSTANCE;
+
+    @Override
+    public boolean apply(@Nonnull Measure input) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Remove this "Nonnull" annotation to honor the overridden method's contract. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2638)